source/linux/io_epoll_context.cpp

```diff
@@ -0,0 +1,485 @@
+      LOGX("epoll_ctl EPOLL_CTL_ADD timerFd_ failed with %i\n", errorCode);
+      throw std::system_error{errorCode, std::system_category()};
+    }
+  }
+
+  {
+    int fd = eventfd(0, EFD_CLOEXEC | EFD_NONBLOCK);
+    // LOGX("eventfd result: %i\n", fd);
+    if (fd < 0) {
+      int errorCode = errno;
+      LOGX("eventfd failed with %i\n", errorCode);
+      throw std::system_error{errorCode, std::system_category()};
+    }
+
+    remoteQueueEventFd_ = safe_file_descriptor{fd};
+  }
+
+  {
+    epoll_event event;
+    event.events = EPOLLIN;
+    event.data.u64 = remote_queue_event_user_data;
+    int result = epoll_ctl(epollFd_.get(), EPOLL_CTL_ADD, remoteQueueEventFd_.get(), &event);
+    // LOGX("epoll_ctl EPOLL_CTL_ADD remoteQueueEventFd_ result: %i\n", result);
+    if (result < 0) {
+      int errorCode = errno;
+      LOGX("epoll_ctl EPOLL_CTL_ADD remoteQueueEventFd_ failed with %i\n", errorCode);
+      throw std::system_error{errorCode, std::system_category()};
+    }
+  }
+
+  LOG("io_epoll_context construction done");
+}
+
+io_epoll_context::~io_epoll_context() {
+  epoll_event event;
+  (void)epoll_ctl(epollFd_.get(), EPOLL_CTL_DEL, remoteQueueEventFd_.get(), &event);
+  (void)epoll_ctl(epollFd_.get(), EPOLL_CTL_DEL, timerFd_.get(), &event);
+  LOG("io_epoll_context destructor done");
+}
+
+void io_epoll_context::run_impl(const bool& shouldStop) {
+  LOG("run loop started");
+
+  auto* oldContext = std::exchange(currentThreadContext, this);
+  scope_guard g = [=]() noexcept {
+    std::exchange(currentThreadContext, oldContext);
+    LOG("run loop exited");
+  };
+
+  while (true) {
+    // Dequeue and process local queue items (ready to run)
+    execute_pending_local();
+
+    if (shouldStop) {
+      break;
+    }
+
+    // Check for any new completion-queue items.
+    acquire_completion_queue_items();
+
+    if (timersAreDirty_) {
+      update_timers();
+    }
+
+    // Check for remotely-queued items.
+    // Only do this if we haven't submitted a poll operation for the
+    // completion queue - in which case we'll just wait until we receive the
+    // completion-queue item).
+    if (!remoteQueueReadSubmitted_) {
+      acquire_remote_queued_items();
+    }
+
+    if (localQueue_.empty()) {
+      if (!remoteQueueReadSubmitted_) {
+        LOG("try_register_remote_queue_notification()");
+        remoteQueueReadSubmitted_ = try_register_remote_queue_notification();
+      }
+
+      LOG("epoll_wait()");
+      int result = epoll_wait(
+        epollFd_.get(),
+        completed_,
+        maxCount_,
+        -1);
+      if (result < 0) {
+        int errorCode = errno;
+        throw std::system_error{errorCode, std::system_category()};
+      }
+      LOGX("epoll_wait() returned %i\n", result);
+      count_ = result;
+    }
+  }
+}
+
+bool io_epoll_context::is_running_on_io_thread() const noexcept {
+  return this == currentThreadContext;
+}
+
+void io_epoll_context::schedule_impl(operation_base* op) {
```

Comment:
This method is not marked `noexcept` but all of the methods it's calling are marked `noexcept`.
Can we mark this one noexcept as well?
Or should we remove the noexcept declaration on schedule_remote()?